### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by gabbar23**
This pull request includes many refactorings that improve the quality and readability of the codebase. Specifically, I have used the following techniques to enhance the code:

- Extract Method: This technique involves breaking down a significant method into smaller, more manageable pieces. Doing so makes the code easier to understand and maintain, and the smaller methods can be reused in other parts of the codebase.

- Rename Method/Variable: This technique involves giving more meaningful names to methods and variables, which helps in understanding the purpose and functionality of the code. By doing so, the code becomes more expressive and easier to comprehend.

- Decompose Conditional: This technique involves breaking down complex conditional statements into smaller, more manageable parts. Doing so makes the code more readable and maintainable, and the more minor conditions can be reused in other parts of the codebase.

- Extract Class: This technique separates a large class into smaller, more specialized classes. Doing so makes the code more modular, and the individual classes can be reused in other parts of the codebase.
 
- Replace Conditional with Polymorphism: This technique replaces complex conditional statements with polymorphic behavior. Doing so makes the code more modular and extensible, and the individual classes can be reused in other parts of the codebase.
 
- Push-Down Variable/Method: This technique involves moving a variable or method from a superclass to a subclass. Doing so makes the code more modular, and the individual classes can be reused in other parts of the codebase.

### CODE REVIEW

File: core/src/main/java/com/digitalpebble/stormcrawler/bolt/FeedParserBolt.java [hunk 1]
```diff
@@ -52,11 +52,11 @@
 /** Extracts URLs from feeds */
 public class FeedParserBolt extends StatusEmitterBolt {
 
-    public static final String isFeedKey = "isFeed";
+    public static final String IS_FEED_METADATA_KEY = "isFeed";
 
     private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(FeedParserBolt.class);
 
-    private boolean sniffWhenNoMDKey = false;
+    private boolean SHOULD_SNIFF_FOR_FEED  = false;
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/bolt/FeedParserBolt.java:59]: This isn't a constant, why do we rename it?

File: core/src/main/java/com/digitalpebble/stormcrawler/filtering/basic/CharacterEscaper.java [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package com.digitalpebble.stormcrawler.filtering.basic;
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/filtering/basic/CharacterEscaper.java:1]: Missing license header

File: core/src/main/java/com/digitalpebble/stormcrawler/filtering/basic/ControlCharacterEscaper.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package com.digitalpebble.stormcrawler.filtering.basic;
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/filtering/basic/ControlCharacterEscaper.java:1]: Missing license header

File: core/src/main/java/com/digitalpebble/stormcrawler/parse/NodeVisitorTraverser.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
+package com.digitalpebble.stormcrawler.parse;
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/parse/NodeVisitorTraverser.java:1]: Missing license header

File: external/elasticsearch/src/main/java/com/digitalpebble/stormcrawler/elasticsearch/persistence/ScrollSpout.java [hunk 5]
```diff
@@ -81,46 +81,52 @@ public void nextTuple() {
         populateBuffer();
     }
 
+    //Rfactored by Amanjot Singh
```
Comment (rzo1) [inline external/elasticsearch/src/main/java/com/digitalpebble/stormcrawler/elasticsearch/persistence/ScrollSpout.java:84]: Should be removed.

File: core/src/main/java/com/digitalpebble/stormcrawler/filtering/basic/ControlCharacterEscaper.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+import java.util.Locale;
+
+public class ControlCharacterEscaper implements CharacterEscaper {
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/filtering/basic/ControlCharacterEscaper.java:5]: This change opens a way for testability. Could you add a JUnit5 test?

File: core/src/main/java/com/digitalpebble/stormcrawler/filtering/basic/NonControlCharacterEscaper.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package com.digitalpebble.stormcrawler.filtering.basic;
+
+public class NonControlCharacterEscaper implements CharacterEscaper {
+    @Override
+    public String escape(byte b) {
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/filtering/basic/NonControlCharacterEscaper.java:5]: This change opens a way for testability. Could you add a JUnit5 test?

File: core/src/main/java/com/digitalpebble/stormcrawler/parse/NodeVisitorTraverser.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+import org.jsoup.select.NodeVisitor;
+
+public class NodeVisitorTraverser {
+    /**
+     //     * Start a depth-first traverse of the root and all of its descendants.
+     //     *
+     //     * @param visitor Node visitor.
+     //     * @param root the root node point to traverse.
+     //     */
+    public static void traverse(
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/parse/NodeVisitorTraverser.java:19]: This change opens a way for testability. Could you add a JUnit5 test?

Comment (rzo1) [reply core/src/main/java/com/digitalpebble/stormcrawler/parse/NodeVisitorTraverser.java:19]: Do we need `static` here? (after we have extracted the class?)

File: core/src/main/java/com/digitalpebble/stormcrawler/parse/NodeVisitorTraverser.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+                    depth--;
+                }
+                visitor.tail(node, depth);
+                if (node == root) break;
+                node = node.nextSibling();
+            }
+        }
+    }
+
+    static void appendNormalisedText(final StringBuilder accum, final TextNode textNode) {
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/parse/NodeVisitorTraverser.java:69]: see above. Perhaps add a JUnit test and remove static?

File: core/src/main/java/com/digitalpebble/stormcrawler/parse/NodeVisitorTraverser.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        final String text = textNode.getWholeText();
+        if (textNode instanceof CDataNode || preserveWhitespace(textNode.parent()))
+            accum.append(text);
+        else StringUtil.appendNormalisedWhitespace(accum, text, lastCharIsWhitespace(accum));
+    }
+        @Contract("null -> false")
+    static boolean preserveWhitespace(Node node) {
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/parse/NodeVisitorTraverser.java:76]: see above. Perhaps add a JUnit test and remove static?

File: external/elasticsearch/src/main/java/com/digitalpebble/stormcrawler/elasticsearch/persistence/ScrollSpout.java [hunk 11]
```diff
@@ -81,46 +81,52 @@ public void nextTuple() {
...
+        LOG.debug("{} ES query {}", logIdprefix, searchRequest.toString());
+    }
+
+    private void doScrollRequest() {
         SearchScrollRequest scrollRequest = new SearchScrollRequest(scrollId);
         scrollRequest.scroll(TimeValue.timeValueMinutes(5L));
 
         isInQuery.set(true);
         client.scrollAsync(scrollRequest, RequestOptions.DEFAULT, this);
-        // dump query to log
```
Comment (rzo1) [inline external/elasticsearch/src/main/java/com/digitalpebble/stormcrawler/elasticsearch/persistence/ScrollSpout.java:120]: any reason for removing the comment?

File: external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/WARCSpout.java [hunk 12]
```diff
@@ -550,4 +550,4 @@ public void declareOutputFields(OutputFieldsDeclarer declarer) {
     public void fail(Object msgId) {
         LOG.error("Failed - unable to replay WARC record of: {}", msgId);
     }
-}
+}
```
Comment (rzo1) [inline external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/WARCSpout.java:553]: Change is unclear to me?

File: core/src/main/java/com/digitalpebble/stormcrawler/bolt/URLPartitionerBolt.java [hunk 13]
```diff
@@ -141,36 +141,30 @@ public void prepare(
...
-                && !mode.equals(Constants.PARTITION_MODE_HOST)) {
+        if (!isValidPartitionMode(mode)) {
             LOG.error("Unknown partition mode : {} - forcing to byHost", mode);
             mode = Constants.PARTITION_MODE_HOST;
         }
 
         LOG.info("Using partition mode : {}", mode);
 
         _collector = collector;
-        // Register a "MultiCountMetric" to count different events in this bolt
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/bolt/URLPartitionerBolt.java:155]: IMHO we should retain the comments

File: core/src/main/java/com/digitalpebble/stormcrawler/bolt/URLPartitionerBolt.java [hunk 14]
```diff
@@ -141,36 +141,30 @@ public void prepare(
...
-        // Storm will emit the counts every n seconds to a special bolt via a
-        // system stream
-        // The data can be accessed by registering a "MetricConsumer" in the
-        // topology
         this.eventCounter = context.registerMetric("URLPartitioner", new MultiCountMetric(), 10);
 
         final int MAX_ENTRIES = 500;
-        cache =
-                new LinkedHashMap(MAX_ENTRIES + 1, .75F, true) {
-                    // This method is called just after a new entry has been added
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/bolt/URLPartitionerBolt.java:165]: IMHO we should retain the comments

File: core/src/main/java/com/digitalpebble/stormcrawler/bolt/URLPartitionerBolt.java [hunk 15]
```diff
@@ -141,36 +141,30 @@ public void prepare(
...
-                    @Override
-                    public boolean removeEldestEntry(Map.Entry eldest) {
-                        return size() > MAX_ENTRIES;
-                    }
-                };
-
-        // If the cache is to be used by multiple threads,
```
Comment (rzo1) [inline core/src/main/java/com/digitalpebble/stormcrawler/bolt/URLPartitionerBolt.java:172]: IMHO we should retain the comments

**Review by rzo1 — CHANGES_REQUESTED** [review]
Can you sign-off your commits (as stated in the README) ? In addition, I added some more comments.

**Comment by jnioche** [comment]
Closing for lack of activity. Please feel free to reopen

**Pull request closed without merging**